Only on Windows: Failed startup of context oeje8w.WebAppContext

**Jetty version(s)**
Jetty 12.0.16

**Jetty Environment**
ee8

**Java version/vendor** `(use: java -version)`
Linux:
openjdk 21.0.5 2024-10-15
OpenJDK Runtime Environment (build 21.0.5+11-Ubuntu-1ubuntu122.04)
OpenJDK 64-Bit Server VM (build 21.0.5+11-Ubuntu-1ubuntu122.04, mixed mode, sharing)

Windows:
openjdk 21.0.5 2024-10-15 LTS
OpenJDK Runtime Environment Temurin-21.0.5+11 (build 21.0.5+11-LTS)
OpenJDK 64-Bit Server Temurin-21.0.5+11 (build 21.0.5+11-LTS, mixed mode, sharing)


**OS type/version**
Linux, 22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Nov 6 17:42:15 UTC 2
Windows 11

**Description**

**In short:**
My simple 'Hello World' WebApp works with Jetty 12.0.16 under Linux. But it does not works with Jetty 12.0.16 under Windows. 

**Details:**

I'm using Jetty in an OSGi Container. I'm using this bundles:
```
jetty-deploy-12.0.16.jar
jetty-ee-12.0.16.jar
jetty-ee8-annotations-12.0.16
jetty-ee8-nested-12.0.16
jetty-ee8-osgi-boot-12.0.16
jetty-ee8-plus-12.0.16
jetty-ee8-security-12.0.16
jetty-ee8-servlet-12.0.16
jetty-ee8-servlets-12.0.16
jetty-ee8-webapp-12.0.16
jetty-http-12.0.16.jar
jetty-io-12.0.16.jar
jetty-jmx-12.0.16.jar
jetty-jndi-12.0.16.jar
jetty-osgi-12.0.16.jar
jetty-plus-12.0.16.jar
jetty-security-12.0.16.jar
jetty-server-12.0.16.jar
jetty-servlet-api-4.0.6
jetty-session-12.0.16.jar
jetty-util-12.0.16.jar
jetty-xml-12.0.16.jar
```
All required bundles like org.osgi.*, org.apache.aries.spifly.dynamic.bundle, ... are installed. 

This are the relevant entries in the MANIFEST.MF:

```
Import-Package: javax.servlet;version="[4.0,5)",javax.servlet.annotation ;version="[4.0,5)",javax.servlet.http;version="[4.0,5)",org.eclipse.jetty.ee8.servlets
Jetty-Environment: ee8
Web-ContextPath: /
```
I'm using a normal WEB-INF/web-xml to configure the servlet. Jetty is configured using jetty.xml, jetty-deploy.xml and jetty-http.xml:

**jetty.xml**
```
<Configure id="Server" class="org.eclipse.jetty.server.Server">

    <!-- =========================================================== -->
    <!-- Server Thread Pool                                          -->
    <!-- =========================================================== -->
    <Get name="ThreadPool">
      <Set name="minThreads">10</Set>
      <Set name="maxThreads">200</Set>
    </Get>


    <!-- =========================================================== -->
    <!-- Set handler Collection Structure                            -->
    <!-- =========================================================== -->
    <Set name="handler">
      <New id="Contexts" class="org.eclipse.jetty.server.handler.ContextHandlerCollection"/>
    </Set>

    <New id="httpConfig" class="org.eclipse.jetty.server.HttpConfiguration">
      <Set name="secureScheme"><Property name="jetty.httpConfig.secureScheme" default="https" /></Set>
      <Set name="securePort"><Property name="jetty.httpConfig.securePort" default="8443" /></Set>
      <Set name="outputBufferSize"><Property name="jetty.httpConfig.outputBufferSize" default="131072" /></Set>
      <Set name="outputAggregationSize"><Property name="jetty.httpConfig.outputAggregationSize" default="131072" /></Set>
      <Set name="requestHeaderSize"><Property name="jetty.httpConfig.requestHeaderSize" default="131072" /></Set>
      <Set name="responseHeaderSize"><Property name="jetty.httpConfig.responseHeaderSize" default="131072" /></Set>
      <Set name="sendServerVersion"><Property name="jetty.httpConfig.sendServerVersion" default="true" /></Set>
      <Set name="sendDateHeader"><Property name="jetty.httpConfig.sendDateHeader" default="false" /></Set>
      <Set name="headerCacheSize"><Property name="jetty.httpConfig.headerCacheSize" default="1024" /></Set>
      <Set name="delayDispatchUntilContent"><Property name="jetty.httpConfig.delayDispatchUntilContent" default="true"/></Set>
      <Set name="maxErrorDispatches"><Property name="jetty.httpConfig.maxErrorDispatches" default="10"/></Set>
      <Set name="persistentConnectionsEnabled"><Property name="jetty.httpConfig.persistentConnectionsEnabled" default="true"/></Set>
      <Set name="requestCookieCompliance"><Call class="org.eclipse.jetty.http.CookieCompliance" name="valueOf"><Arg><Property name="jetty.httpConfig.requestCookieCompliance" default="RFC6265"/></Arg></Call></Set>
      <Set name="responseCookieCompliance"><Call class="org.eclipse.jetty.http.CookieCompliance" name="valueOf"><Arg><Property name="jetty.httpConfig.responseCookieCompliance" default="RFC6265"/></Arg></Call></Set>
    </New>

    <!-- =========================================================== -->
    <!-- extra options                                               -->
    <!-- =========================================================== -->
    <Set name="stopAtShutdown"><Property name="jetty.server.stopAtShutdown" default="true"/></Set>
    <Set name="stopTimeout"><Property name="jetty.server.stopTimeout" default="5000"/></Set>
    <Set name="dumpAfterStart"><Property name="jetty.server.dumpAfterStart" default="false"/></Set>
    <Set name="dumpBeforeStop"><Property name="jetty.server.dumpBeforeStop" default="false"/></Set>


    <Call class="java.lang.System" name="setProperty">
      <Arg>java.naming.factory.initial</Arg>
      <Arg><Property name="java.naming.factory.initial" default="org.eclipse.jetty.jndi.InitialContextFactory"/></Arg>
    </Call>
    <Call class="java.lang.System" name="setProperty">
      <Arg>java.naming.factory.url.pkgs</Arg>
      <Arg><Property name="java.naming.factory.url.pkgs" default="org.eclipse.jetty.jndi"/></Arg>
    </Call>

</Configure>

```

**jetty-deploy.xml**
```
<?xml version="1.0"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "https://eclipse.dev/jetty/configure_9_3.dtd">

<Configure id="Server" class="org.eclipse.jetty.server.Server">

    <!-- =========================================================== -->
    <!-- Configure the deployment manager                            -->
    <!-- =========================================================== -->
    <Call name="addBean">
      <Arg>
        <New id="DeploymentManager" class="org.eclipse.jetty.deploy.DeploymentManager">
          <Set name="contexts">
            <Ref refid="Contexts" />
          </Set>
        </New>
      </Arg>
    </Call>

</Configure>

```
**jetty-http.xml**
```
<?xml version="1.0"?>
<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "https://eclipse.dev/jetty/configure_9_3.dtd">

<!-- ============================================================= -->
<!-- Configure the Jetty Server instance with an ID "Server"       -->
<!-- by adding an HTTP connector.                                   -->
<!-- This configuration must be used in conjunction with jetty.xml -->
<!-- ============================================================= -->
<Configure id="Server" class="org.eclipse.jetty.server.Server">

  <!-- =========================================================== -->
  <!-- Add an HTTP Connector.                                       -->
  <!-- Configure an o.e.j.server.ServerConnector with a single     -->
  <!-- HttpConnectionFactory instance using the common httpConfig  -->
  <!-- instance defined in jetty.xml                               -->
  <!--                                                             -->
  <!-- Consult the javadoc of o.e.j.server.ServerConnector and     -->
  <!-- o.e.j.server.HttpConnectionFactory for all configuration    -->
  <!-- that may be set here.                                       -->
  <!-- =========================================================== -->
  <Call name="addConnector">
    <Arg>
      <New class="org.eclipse.jetty.server.ServerConnector">
        <Arg name="server"><Ref refid="Server" /></Arg>
        <Arg name="factories">
          <Array type="org.eclipse.jetty.server.ConnectionFactory">
            <Item>
              <New class="org.eclipse.jetty.server.HttpConnectionFactory">
                <Arg name="config"><Ref refid="httpConfig" /></Arg>
              </New>
            </Item>
          </Array>
        </Arg>
        <Set name="host"><Property name="jetty.http.host" /></Set>
        <Set name="port"><Property name="jetty.http.port" default="8080" /></Set>
        <Set name="idleTimeout"><Property name="jetty.http.idleTimeout" default="30000"/></Set>
      </New>
    </Arg>
  </Call>

</Configure>
```

This works with 
- 12.0.16 / Linux

It does not work with:
- jetty 12.0.16 / Windows

The error message is

```
WARNING: Failed startup of context oeje8w.WebAppContext@1850f359{/test.product,/test.product,null,false}{file:/X:/test-production/test-server/vsrv-test/test.server.test/vsrv/server/test-server-data/org.eclipse.osgi/173/0/bundleFile} [Fri Jan 17 13:40:08 CET 2025]
```
Calling the servlet return a 503 (not 404!)

@joakime @janbartel Sorry, the last point is wrong. When running your test cases the setWarResource method is called!

Until now I have used Linux to analyze this issue. Today I have build and test this under Windows. Running
````
mvn test
````
in `jetty-ee9\jetty-ee9-osgi\test-jetty-ee9-osgi` yields to the same error (503 - service not available) as in my test environment. e.g.:

````
[ERROR]   TestJettyOSGiBootWithAnnotations.testIndex:100 Response status code expected:<200> but was:<503>
[ERROR]   TestJettyOSGiBootWithJsp.testJspDump:84 expected:<200> but was:<503>
[ERROR]   TestJettyOSGiClasspathResources.testWebInfResourceNotOnBundleClasspath:94 expected:<200> but was:<503>
````


Under Linux all tests passed

Hi @janbartel ,
I have tried to build your branch, but it fails with:
```
  [ERROR] Failed to execute goal org.apache.maven.plugins:maven-dependency-plugin:3.8.1:copy (copy) on project jetty-home: Unable to find/resolve artifact.: The following artifacts could not be resolved: org.eclipse.jetty:jetty-project:txt:version:12.0.17-SNAPSHOT (absent): org.eclipse.jetty:jetty-project:txt:version:12.0.17-SNAPSHOT was not found
```
Currently I'm busy. Will take a look at this later.

Best regards
Gregor

--- a/jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/OSGiApp.java
+++ b/jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/OSGiApp.java
@@ -29,6 +29,7 @@ import org.eclipse.jetty.osgi.util.BundleFileLocatorHelperFactory;
 import org.eclipse.jetty.server.Deployable;
 import org.eclipse.jetty.server.handler.ContextHandler;
 import org.eclipse.jetty.util.StringUtil;
+import org.eclipse.jetty.util.URIUtil;
 import org.eclipse.jetty.util.resource.Resource;
 import org.eclipse.jetty.util.resource.ResourceFactory;
 import org.osgi.framework.Bundle;
@@ -80,7 +81,7 @@ public class OSGiApp extends App
         File bundleLocation = BundleFileLocatorHelperFactory.getFactory().getHelper().getBundleInstallLocation(bundle);
         File root = (bundleOverrideLocation == null ? bundleLocation : new File(bundleOverrideLocation));
         //Fix some osgiPaths.get( locations which point to an archive, but that doesn't end in .jar 
-        URL url = BundleFileLocatorHelperFactory.getFactory().getHelper().getLocalURL(root.toURI().toURL());
+        URL url = BundleFileLocatorHelperFactory.getFactory().getHelper().getLocalURL(URIUtil.correctURI(root.toURI()).toURL());
         
         return ResourceFactory.root().newResource(url);
     }

--- a/jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java
+++ b/jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/util/Util.java
@@ -29,6 +29,7 @@ import java.util.StringTokenizer;
 
 import org.eclipse.jetty.osgi.OSGiServerConstants;
 import org.eclipse.jetty.util.StringUtil;
+import org.eclipse.jetty.util.URIUtil;
 import org.osgi.framework.Bundle;
 import org.osgi.framework.BundleContext;
 import org.osgi.framework.Filter;
@@ -61,7 +62,7 @@ public class Util
             return null;
 
         if (path.startsWith("file:/"))
-            return new URI(path);
+            return URIUtil.correctURI(new URI(path));
 
         if (path.startsWith("/") && File.separatorChar != '/')
             return new URI("file:" + path);

--- a/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebInfConfiguration.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebInfConfiguration.java
@@ -207,17 +207,14 @@ public class WebInfConfiguration extends AbstractConfiguration
                 // Look for sibling directory.
                 Path extractedWebAppDir = null;
 
-                if (war != null)
-                {
-                    Path warPath = context.getResourceFactory().newResource(war).getPath();
+                Path warPath = webApp.getPath();
 
-                    // look for a sibling like "foo/" to a "foo.war"
-                    if (warPath != null && FileID.isWebArchive(warPath) && Files.exists(warPath))
-                    {
-                        Path sibling = warPath.getParent().resolve(FileID.getBasename(warPath));
-                        if (Files.exists(sibling) && Files.isDirectory(sibling) && Files.isWritable(sibling))
-                            extractedWebAppDir = sibling;
-                    }
+                // look for a sibling like "foo/" to a "foo.war"
+                if (warPath != null && FileID.isWebArchive(warPath) && Files.exists(warPath))
+                {
+                    Path sibling = warPath.getParent().resolve(FileID.getBasename(warPath));
+                    if (Files.exists(sibling) && Files.isDirectory(sibling) && Files.isWritable(sibling))
+                        extractedWebAppDir = sibling;
                 }
 
                 if (extractedWebAppDir == null)

--- a/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebInfConfiguration.java
+++ b/jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/WebInfConfiguration.java
@@ -204,17 +204,25 @@ public class WebInfConfiguration extends AbstractConfiguration
                     !webApp.isDirectory()
             )
             {
-                // Look for sibling directory.
                 Path extractedWebAppDir = null;
 
-                Path warPath = webApp.getPath();
-
-                // look for a sibling like "foo/" to a "foo.war"
-                if (warPath != null && FileID.isWebArchive(warPath) && Files.exists(warPath))
+                // If this is a war file, we should look for a sibling
+                // directory of the same name where it has been already extracted
+                if (war != null)
                 {
-                    Path sibling = warPath.getParent().resolve(FileID.getBasename(warPath));
-                    if (Files.exists(sibling) && Files.isDirectory(sibling) && Files.isWritable(sibling))
-                        extractedWebAppDir = sibling;
+                    // We have obtained the webApp from the war string, so it
+                    // cannot be a CombinedResource, therefore safe to use it's Path
+                    Path warPath = webApp.getPath();
+                    if (warPath != null)
+                    {
+                        // look for a sibling like "foo/" to a "foo.war"
+                        if (FileID.isWebArchive(warPath) && Files.exists(warPath))
+                        {
+                            Path sibling = warPath.getParent().resolve(FileID.getBasename(warPath));
+                            if (Files.exists(sibling) && Files.isDirectory(sibling) && Files.isWritable(sibling))
+                                extractedWebAppDir = sibling;
+                        }
+                    }
                 }
 
                 if (extractedWebAppDir == null)

--- a/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
@@ -212,10 +212,9 @@ public class WebInfConfiguration extends AbstractConfiguration
 
                 if (war != null)
                 {
-                    Resource warResource = context.getResourceFactory().newResource(war);
-                    Path warPath = warResource.getPath();
+                    Path warPath = webApp.getPath();
                     if (warPath == null)
-                        throw new IllegalStateException("No path for " + warResource);
+                        throw new IllegalStateException("No path for " + webApp);
 
                     // look for a sibling like "foo/" to a "foo.war"
                     if (FileID.isWebArchive(warPath) && Files.exists(warPath))

--- a/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
@@ -209,13 +209,9 @@ public class WebInfConfiguration extends AbstractConfiguration
             {
                 // Look for sibling directory.
                 Path extractedWebAppDir = null;
-
-                if (war != null)
+                Path warPath = webApp.getPath();
+                if (warPath != null)
                 {
-                    Path warPath = webApp.getPath();
-                    if (warPath == null)
-                        throw new IllegalStateException("No path for " + webApp);
-
                     // look for a sibling like "foo/" to a "foo.war"
                     if (FileID.isWebArchive(warPath) && Files.exists(warPath))
                     {

--- a/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
@@ -212,7 +212,11 @@ public class WebInfConfiguration extends AbstractConfiguration
 
                 if (war != null)
                 {
-                    Path warPath = Path.of(war);
+                    Resource warResource = context.getResourceFactory().newResource(war);
+                    Path warPath = warResource.getPath();
+                    if (warPath == null)
+                        throw new IllegalStateException("No path for " + warResource);
+
                     // look for a sibling like "foo/" to a "foo.war"
                     if (FileID.isWebArchive(warPath) && Files.exists(warPath))
                     {

--- a/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
+++ b/jetty-ee9/jetty-ee9-webapp/src/main/java/org/eclipse/jetty/ee9/webapp/WebInfConfiguration.java
@@ -207,17 +207,24 @@ public class WebInfConfiguration extends AbstractConfiguration
                     !webApp.isDirectory())
             )
             {
-                // Look for sibling directory.
                 Path extractedWebAppDir = null;
-                Path warPath = webApp.getPath();
-                if (warPath != null)
+
+                // If this is a war file, we should look for a sibling
+                // directory of the same name
+                if (war != null)
                 {
-                    // look for a sibling like "foo/" to a "foo.war"
-                    if (FileID.isWebArchive(warPath) && Files.exists(warPath))
+                    // We have obtained the webApp from the war string, so it
+                    // cannot be a CombinedResource, therefore safe to use it's Path
+                    Path warPath = webApp.getPath();
+                    if (warPath != null)
                     {
-                        Path sibling = warPath.getParent().resolve(FileID.getBasename(warPath));
-                        if (Files.exists(sibling) && Files.isDirectory(sibling) && Files.isWritable(sibling))
-                            extractedWebAppDir = sibling;
+                        // look for a sibling like "foo/" to a "foo.war"
+                        if (FileID.isWebArchive(warPath) && Files.exists(warPath))
+                        {
+                            Path sibling = warPath.getParent().resolve(FileID.getBasename(warPath));
+                            if (Files.exists(sibling) && Files.isDirectory(sibling) && Files.isWritable(sibling))
+                                extractedWebAppDir = sibling;
+                        }
                     }
                 }
 

